src/main/java/io/usethesource/vallang/impl/util/collections/ShareableValuesHashSet.java

```diff
@@ -454,55 +335,62 @@ public String toString(){
 	}
 	
 	private static class SetIterator implements Iterator<IValue>{
-		private final Entry<IValue>[] data;
+		private final @Nullable Entry<IValue>[] data;
 		
-		private Entry<IValue> current;
+		private @Nullable Entry<IValue> current;
 		private int index;
 		
-		public SetIterator(Entry<IValue>[] entries){
+		public SetIterator(@Nullable Entry<IValue>[] entries) {
 			super();
 			
 			data = entries;
 
 			index = data.length - 1;
-			current = new Entry<>(0, null, data[index]);
-			locateNext();
+			current = new Entry<>(0, ValueFactory.getInstance().bool(true) /* dummy */, data[index]);
+			locateNext(data);
 		}
 		
-		private void locateNext(){
-			Entry<IValue> next = current.next;
-			if(next != null){
-				current = next;
+		private void locateNext(@UnknownInitialization SetIterator this, @Nullable Entry<IValue>[] data) {
+			Entry<IValue> next = current != null ? current.next : null;
+
+			if (next != null) {
+				this.current = next;
 				return;
 			}
 			
-			for(int i = index - 1; i >= 0 ; i--){
-				Entry<IValue> entry = data[i];
-				if(entry != null){
-					current = entry;
-					index = i;
+			for (int i = this.index - 1; i >= 0 ; i--) {
+				@Nullable Entry<IValue> entry = data[i];
+				if (entry != null) {
+					this.current = entry;
+					this.index = i;
 					return;
 				}
 			}
 			
-			current = null;
-			index = 0;
+			this.current = null;
+			this.index = 0;
 		}
 		
-		public boolean hasNext(){
+		@EnsuresNonNullIf(expression="this.current", result=true)
+		@Override
+		public boolean hasNext(@UnknownInitialization SetIterator this) {
 			return (current != null);
```

Comment:
NP: to keep up with the style of the previous function, shouldn't this also be `this.current`?